src/ripple/protocol/impl/LedgerFormats.cpp

```diff
@@ -271,13 +271,25 @@ LedgerFormats::LedgerFormats()
     add(jss::AMM,
         ltAMM,
         {
-            {sfAccount, soeREQUIRED},
-            {sfTradingFee, soeDEFAULT},
-            {sfVoteSlots, soeOPTIONAL},
-            {sfAuctionSlot, soeOPTIONAL},
-            {sfLPTokenBalance, soeREQUIRED},
-            {sfAsset, soeREQUIRED},
-            {sfAsset2, soeREQUIRED}
+            {sfAccount,              soeREQUIRED},
+            {sfTradingFee,           soeDEFAULT},
+            {sfVoteSlots,            soeOPTIONAL},
+            {sfAuctionSlot,          soeOPTIONAL},
+            {sfLPTokenBalance,       soeREQUIRED},
+            {sfAsset,                soeREQUIRED},
+            {sfAsset2,               soeREQUIRED}
+        },
+        commonFields);
+
+    add(jss::DID,
+        ltDID,
+        {
+            {sfAccount,              soeREQUIRED},
+            {sfURI,                  soeOPTIONAL},
+            {sfData,                 soeOPTIONAL},
+            {sfOwnerNode,            soeREQUIRED},
+            {sfPreviousTxnID,        soeREQUIRED},
+            {sfPreviousTxnLgrSeq,    soeREQUIRED}
```

Comment:
I feel like `DID` is a weird name for this object. This is just a ledger object that holds two arbitrary bits of data, in the `URI` field and the `Data` field. Why not call this something more generic? I could see this being used for things other than DID.

Reply:
You can only own one of them, so I don't really know what else you'd use them for.

Reply:
This is just arbitrary data storage, I could really use this for anything at all. I could use it for on oracle, a bridge, governance, etc.

Reply:
That's an interesting point @cjcobb23 

We could make this more generic - add a read-only "kind" field that's part of the hash and let people store whatever data they want.

So a DID would be stored in this new generic data object and have a "kind" of "DID", oracles would have a "kind" of "oracle-usd-xrp" or somesuch, and since the "kind" is part of the hash they are distinct objects. Although with this design you can only ever have one object per "kind", but that doesn't seem too bad.

We have talked about general purpose data storage in the past, now that we have a concrete use-case for such an object (DIDs) it seems like it's worth looking at doing the generic solution. We'd have to flesh it out a little more if we want to go in this direction, but it seems like a good idea.

I'd probably rename the fields, maybe add a couple more fields, and a read-only "kind" and everything else works as-is now. It's an easy change from an implementation POV.

One problem with the generic solution is getting everyone to agree on what "kind" means. I might want ids to be "IDS", you might want "DID", and someone else whatever the french word for "ID" is. But that doesn't seem like a killer.

Reply:
>    Although with this design you can only ever have one object per "kind", but that doesn't seem too bad.

@seelabs there could be multiple oracles even for the same type. I.e. could have a pricing oracle for XRP/USD by chainlink, DIA, etc. Also could have weather, sports, etc. oracles say from the same Oracle Provider.

Reply:
@gregtatcam If the oracles are from different providers they can be on different accounts (the hash would include the "kind"and the account). Weather, sports, ect oracles from the same providers could be different "kinds".

Reply:
Could also add a "description" field as well, that's just some plain text description of the data.

One thing I think worth thinking about is actually storing the schema in the object. The schema could be defined in protobuf. This could enable things like allowing callers to pass in protobuf in the update transaction, safely deserializing the protobuf to other types and returning JSON to the RPC callers by just converting from protobuf to JSON. Obviously this wouldn't work for very large schemas, but for small schemas, say with just a few fields, we could probably store them in the object and get a lot of benefits.

Reply:
@mDuo13 What are the hazards of a large ledger entry object? I'm not familiar with the issues of trust line creation.

The owner reserve fees disincentives spamming an account with frivolous DIDs right?

Reply:
https://github.com/XRPLF/XRPL-Standards/discussions/132